sys/rtt_stdio: Support SEGGER RTT for stdin/stdout

immesys opened the pull request:
This is a rebase / reclean of #4720 which I accidentally nuked while cleaning stuff up.

@thomaseichinger I addressed the concerns you had on the previous PR. Would you mind taking a look and seeing if I missed anything?

I'd like to get this into the upcoming release, what can I do to increase the odds of that happening?

thomaseichinger commented on sys/rtt_stdio/rtt_stdio.c:
```diff
@@ -0,0 +1,287 @@
+/**
+ * @brief read bytes from the down buffer. This function does not block.
+ *        The logic here is unmodified from SEGGER's reference, it is just
+ *        refactored to match code style. The string is not null terminated.
```
Since we're importing logic unchanged here, was there any license with the SEGGER code?

thomaseichinger commented:
Great to see that one updated. The code is in very good shape, I can not really test however as I'm lacking the setup but I guess you tested it often enough.

I'd appreciate if you could clarify on the license topic. I'm happy to ACK this as soon as this is sorted out.

immesys commented:
I downloaded the original SEGGER implementation again and checked through it and also checked all over the website but I see no mention of a license anywhere. I've dropped them an email, I'll see what they say.

immesys commented:
As for testing, yeah we use this every day, so it's been fairly well debugged.

immesys commented:
Wait, I found the license terms. I'll add this in to the files when I get to work. Thanks for thinking of the legal  stuffs @thomaseichinger.

```
**********************************************************************
*                                                                    *
* All rights reserved.                                               *
*                                                                    *
* SEGGER strongly recommends to not make any changes                 *
* to or modify the source code of this software in order to stay     *
* compatible with the RTT protocol and J-Link.                       *
*                                                                    *
* Redistribution and use in source and binary forms, with or         *
* without modification, are permitted provided that the following    *
* conditions are met:                                                *
*                                                                    *
* o Redistributions of source code must retain the above copyright   *
*   notice, this list of conditions and the following disclaimer.    *
*                                                                    *
* o Redistributions in binary form must reproduce the above          *
*   copyright notice, this list of conditions and the following      *
*   disclaimer in the documentation and/or other materials provided  *
*   with the distribution.                                           *
*                                                                    *
* o Neither the name of SEGGER Microcontroller GmbH & Co. KG         *
*   nor the names of its contributors may be used to endorse or      *
*   promote products derived from this software without specific     *
*   prior written permission.                                        *
*                                                                    *
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND             *
* CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,        *
* INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF           *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE           *
* DISCLAIMED. IN NO EVENT SHALL SEGGER Microcontroller BE LIABLE FOR *
* ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR           *
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT  *
* OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;    *
* OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF      *
* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT          *
* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE  *
* USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH   *
* DAMAGE.                                                            *
*                                                                    *
**********************************************************************
```

immesys commented:
Alright I have added in the original license stuff. I think this is ready.

immesys commented:
meh, I think I did the auto-squash thing wrong. Let me know if you need me to manually squash this (I can't remember if github does that automatically)

immesys commented:
@thomaseichinger ping, do you need anything more from me on this?

thomaseichinger commented:
@immesys I'm sorry, let my try to run some test on boards without it this feature. Hopefully today.

immesys commented:
Thanks!

thomaseichinger commented:
@immesys I'm very sorry it took so long but we should get it in by today.

kYc0o commented:
I'm wondering if the dependency on xtimer would be good for some platforms, like the cc2650stk. I'm doing some work on this platform now and this PR sounds awesome, since I don't have the DevKit to access STDIO. However, xtimer support for cc2650 (#5361) is not working afaik...

immesys commented:
Interestingly, those two useless assignments are present upstream. I guess they put them there as some kind of documentation? Either that or they used to have that code in a loop but since removed the loop. I'll remove it.

immesys commented:
@thomaseichinger I removed those two lines, and fixed the whitespace error. I also squashed for you.

I found the whitespace tool but may I ask what tool found the useless assignment? I'd like to run all these checks on my PRs beforehand in future.

thomaseichinger commented:
@immesys The tool you are looking for should be [cppcheck](http://cppcheck.sourceforge.net) wrapped in `dist/tools/cppcheck/check.sh`.

Thanks for fixing the static analysis messages. Regarding squashing, best practice it usually is to wait until the maintainer says so since it's easier to follow changes and fixes. It's ok for this case now though. 😉

thomaseichinger commented:
@kYc0o I think the `xtimer` dependency is legit here. Do you object or can we merge this?

thomaseichinger commented:
ACK & Go

immesys commented:
Thanks all!
